libtransmission/variant-json.cc

```diff
@@ -569,41 +569,36 @@ void jsonRealFunc(tr_variant const& /*var*/, double const val, void* vdata)
     jsonChildFunc(data);
 }
 
-[[nodiscard]] char* write_escaped_char(char* buf, char const* const end, std::string_view& sv)
-{
-    auto u16buf = std::array<std::uint16_t, 2>{};
-
-    auto const* const begin8 = std::data(sv);
-    auto const* const end8 = begin8 + std::size(sv);
-    auto const* walk8 = begin8;
-    utf8::next(walk8, end8);
-    auto const end16 = utf8::utf8to16(begin8, walk8, std::begin(u16buf));
-
-    for (auto it = std::cbegin(u16buf); it != end16; ++it)
-    {
-        buf = fmt::format_to_n(buf, end - buf - 1, FMT_COMPILE("\\u{:04x}"), *it).out;
-    }
-
-    sv.remove_prefix(walk8 - begin8 - 1);
-    return buf;
-}
-
+// https://datatracker.ietf.org/doc/html/rfc8259#section-7
 void jsonStringFunc(tr_variant const& /*var*/, std::string_view sv, void* vdata)
 {
     auto* const data = static_cast<struct JsonWalk*>(vdata);
 
+    auto const utf8_str = tr_strv_convert_utf8(sv);
+    auto utf8_sv = std::string_view{ utf8_str };
+
     auto& out = data->out;
-    auto const [buf, buflen] = out.reserve_space(std::size(sv) * 6 + 2);
+    auto const [buf, buflen] = out.reserve_space(std::size(utf8_sv) * 6 + 2);
     auto* walk = reinterpret_cast<char*>(buf);
     auto const* const begin = walk;
     auto const* const end = begin + buflen;
 
     *walk++ = '"';
 
-    for (; !std::empty(sv); sv.remove_prefix(1))
+    for (; !std::empty(utf8_sv); utf8_sv.remove_prefix(1))
     {
-        switch (sv.front())
+        switch (utf8_sv.front())
         {
+        case '\\':
+            *walk++ = '\\';
+            *walk++ = '\\';
+            break;
+
+        case '"':
```

Comment:
nit: that reordering is not needed for the fix.
But if you're trying to have it in ASCII order of the resulting string, then `"` should be before `\`, while if you're trying to have in in ASCII order of the escape sequence, then it should be `\b`, `\t`, `\n`, `\f`, `\r`.

Reply:
The list of valid escape sequences was listed in this order in the RFC document.

Reply:
```
              %x22 /          ; "    quotation mark  U+0022
              %x5C /          ; \    reverse solidus U+005C
              %x2F /          ; /    solidus         U+002F
              %x62 /          ; b    backspace       U+0008
              %x66 /          ; f    form feed       U+000C
              %x6E /          ; n    line feed       U+000A
              %x72 /          ; r    carriage return U+000D
              %x74 /          ; t    tab             U+0009
```
Straight from the RFC: https://datatracker.ietf.org/doc/html/rfc8259

Reply:
Damn I must've been blind.

Reply:
deee3f3